### PR TITLE
[RPC] Report error targets and locations

### PR DESCRIPTION
**Pull request by cwong-ocaml**
Properly report the new information as changed in #4513

### CODE REVIEW

**Comment by cwong-ocaml** [comment]
This was failing CI on my personal branch, but I'm not sure why. My best guess is that I missed a catch-reraise of `User_error.E` somewhere, but I'm not sure how to go about debugging that other than grepping for `User_error.E`.

File: src/dune_engine/process.ml [hunk 1]
```diff
@@ -364,9 +364,12 @@ module Exit_status = struct
 
   (* In this module, we don't need the "Error: " prefix given that it is already
      included in the error message from the command. *)
-  let fail paragraphs = raise (User_error.E (User_message.make paragraphs))
+  let fail ~dir paragraphs =
+    match dir with
```
Comment (rgrinberg) [inline src/dune_engine/process.ml:368]: I think it's counter intuitive that we don't report the directory just because we inherited it instead of setting it explicitly.

File: src/dune_util/located_error.mli [hunk 2]
```diff
@@ -0,0 +1,14 @@
+(** A [Located_error.t] is morally a [User_error] annotated with an associated
+    directory. Namely, if a build action raises an error, we want to preserve
+    the directory that action was raised from. *)
+
+open Stdune
+
+exception E of User_message.t * Path.t
```
Comment (rgrinberg) [inline src/dune_util/located_error.mli:7]: This exception subtly changes how we handle user errors elsewhere in the code. See `expander.ml` where we treat user errors in a special way. More generally, it just seems fragile to me to extend the concept of a user error this way. Every time we add a new type of user error, we need to scan the entire code base to make sure we're consistently handling it without help from the type system.

I think we should consider alternatives such as:

- Optional dir argument in `User_error.E`.
- Use open types for the payload of user errors. E.g.

```ocaml
type error = ..
type error += User_error of Pp.t
type error += Located_error of Pp.t * Path.t
exception User_error of error
```

File: src/dune_engine/process.ml (hunk 1 above)
Comment (cwong-ocaml) [reply src/dune_engine/process.ml:368]: I'm not sure I understand, can you elaborate?

File: src/dune_util/located_error.mli (hunk 2 above)
Comment (cwong-ocaml) [reply src/dune_util/located_error.mli:7]: I agree that it's pretty fragile to extend it this way. Personally, I'd rather just have a closed-world of errors (instead of the open variant or adding new exceptions), so the compiler can tell us where we're missing handling.

Directly referencing `Path.t` in `User_error` seems difficult, because `stdune-unstable/path.ml` needs to raise these errors in several cases, and from a brief scan it doesn't seem like it's going to be easy to split out a `Path0` type module either.

File: src/dune_engine/process.ml (hunk 1 above)
Comment (rgrinberg) [reply src/dune_engine/process.ml:368]: The directory is

- `None` whenever we it to remain as cwd
- `Some path` whenever we explicitly set it to something.

So it means that errors from running external commands report the directory only when it is explicitly set. This does not  result in predictable error reporting for clients.

Comment (cwong-ocaml) [reply src/dune_engine/process.ml:368]: Ah, yeah, that seems like it could be a problem. Maybe a better place to pass it would be via `Action_exec`?

Comment (rgrinberg) [reply src/dune_engine/process.ml:368]: You could also just set dir to cwd if it's None when building the exceptions.

**Comment by cwong-ocaml** [comment]
I spoke with Jeremie about this design and he convinced me to reconsider the original design with `Located_error`. As far as we could find, there are only two places where we catch `User_error.E` with the intent of unpacking the information within it -- `Report_error.get_user_message` and `Build_system.Error.info` (formerly `Build_system.Error.message`), so those are the only two cases where we need to catch the new variant. @rgrinberg can you elaborate on what you were referring to with regard to `expander.ml`? I don't see anywhere that we catch a `User_error.E` at all, only raise it.

On a second note, it looks like `Build_system.Error.info` (`message`) is redundant with `Report_error.get_user_message`. If we consolidate the two, we can reduce it to a single place where new error variants should be handled.

**Comment by jeremiedimino** [comment]
Just a nit-pick on the exception name: "located error" doesn't make a lot of sense to me as `User_message.t` already has a location field. The location we are talking about here is the cwd of the program that produced the error message. Additionally, if you want to throw promotions in there in the future, it really feels like we are talking about more specific "action errors" rather general errors.

**Comment by cwong-ocaml** [comment]
I've made the change locally (it's basically just a revert of my last stuff anyway) but would appreciate @rgrinberg 's take before I submit it properly (and un-draft this).

**Comment by rgrinberg** [comment]
> can you elaborate on what you were referring to with regard to expander.ml? I don't see anywhere that we catch a User_error.E at all, only raise it.

Here's the snippet in expander.ml:
```ocaml
(* Make sure to delay exceptions *)
let expand_pform_gen ~context ~bindings ~dir ~source pform =
  match expand_pform_gen ~context ~bindings ~source ~dir pform with
  | exception (User_error.E _ as exn) ->
```

**Comment by rgrinberg** [comment]
The new approach in this PR looks solid to me. Perhaps we should wrap the open type with a functor that introduces a new constructor? This will requier new annotations to provide a `to_dyn` converter. Should be useful when printing these errors.

File: src/dune_engine/process.ml [hunk 3]
```diff
@@ -378,7 +378,9 @@ module Exit_status = struct
       | None -> Path.of_string (Sys.getcwd ())
       | Some dir -> dir
     in
-    User_error.raise paragraphs ~annot:(With_directory_annot.make dir)
+    raise
```
Comment (rgrinberg) [inline src/dune_engine/process.ml:381]: Why was this changed?

Comment (cwong-ocaml) [reply src/dune_engine/process.ml:381]: On main, it was originally `raise`. I had initially changed it to `User_error.raise`, but that was leading to the CI failures where `Error:` got prepended to things, so I reverted the change.

Comment (cwong-ocaml) [reply src/dune_engine/process.ml:381]: I do think that we should standardize the way that we raise user errors (either via `User_error.raise` or raising `User_error.E`), but that seems outside the scope of this PR.

**Comment by rgrinberg** [comment]
> I do think that we should standardize the way that we raise user errors (either via `User_error.raise` or raising `User_error.E`), but that seems outside the scope of this PR.

Sounds good to me.

LGTM.

**Review by rgrinberg — APPROVED** [review]

File: otherlibs/stdune-unstable/user_error.mli [hunk 4]
```diff
@@ -1,15 +1,37 @@
...
 
+module Annot : sig
+  type t
+
+  module type S = sig
+    type payload
+
+    val make : payload -> t
+
+    val check : t -> (payload -> 'a) -> (unit -> 'a) -> 'a
```
Comment (jeremiedimino) [inline otherlibs/stdune-unstable/user_error.mli:11]: We do something very similar in `Dune_util.Persistent`, but IMO using a clearer API. The idea is to use a  "private extensible type":

```
module Annot : sig
  type t = private ..

  module type S = sig
    type payload
    type t += T of payload

    val make : payload -> t
  end
```

With this API, the user can simply pattern match `t` values but still cannot add new constructors to `t` without going through `Make`.
